packages/core/src/browser/secondary-window-handler.ts

```diff
@@ -0,0 +1,249 @@
+
+import debounce = require('lodash.debounce');
+import { inject, injectable } from 'inversify';
+import { BoxLayout, BoxPanel, ExtractableWidget, Widget } from './widgets';
+import { MessageService } from '../common/message-service';
+import { ApplicationShell } from './shell/application-shell';
+import { Emitter } from '../common/event';
+import { WindowService } from './window/window-service';
+
+/** Widget to be contained directly in a secondary window. */
+class SecondaryWindowRootWidget extends Widget {
+
+    constructor() {
+        super();
+        this.layout = new BoxLayout();
+    }
+
+    addWidget(widget: Widget): void {
+        (this.layout as BoxLayout).addWidget(widget);
+        BoxPanel.setStretch(widget, 1);
+    }
+
+}
+
+/**
+ * Offers functionality to move a widget out of the main window to a newly created window.
+ * Widgets must explicitly implement the `ExtractableWidget` interface to support this.
+ *
+ * This handler manages the opened secondary windows and sets up messaging between them and the Theia main window.
+ * In addition, it provides access to the extracted widgets and provides notifications when widgets are added to or removed from this handler.
+ *
+ * _Note:_ This handler is used by the application shell and there should be no need for callers to directly interact with this class.
+ * Instead, consider using the application shell.
+ */
+@injectable()
+export class SecondaryWindowHandler {
+    /** List of currently open secondary windows. Window references should be removed once the window is closed. */
+    protected readonly secondaryWindows: Window[] = [];
+    /** List of widgets in secondary windows. */
+    protected readonly _widgets: ExtractableWidget[] = [];
+
+    /**
+     * Randomized prefix to be included in opened windows' ids.
+     * This avoids conflicts when creating sub-windows from multiple theia instances (e.g. by opening Theia multiple times in the same browser)
+     */
+    protected readonly prefix = crypto.getRandomValues(new Uint16Array(1))[0];
+    /** Unique id. Increase after every access. */
+    private nextId = 0;
+
+    protected applicationShell: ApplicationShell;
+
+    protected readonly onDidAddWidgetEmitter = new Emitter<Widget>();
+    /** Subscribe to get notified when a widget is added to this handler, i.e. the widget was moved to an secondary window . */
+    readonly onDidAddWidget = this.onDidAddWidgetEmitter.event;
+
+    protected readonly onDidRemoveWidgetEmitter = new Emitter<Widget>();
+    /** Subscribe to get notified when a widget is removed from this handler, i.e. the widget's window was closed or the widget was disposed. */
+    readonly onDidRemoveWidget = this.onDidRemoveWidgetEmitter.event;
+
+    constructor(
+        @inject(MessageService) protected readonly messageService: MessageService,
+        @inject(WindowService) protected readonly windowService: WindowService
+    ) { }
```

Comment:
I believe we should use property injection here?

- https://github.com/eclipse-theia/theia/wiki/Coding-Guidelines#property-injection